### DISCUSSION TITLE
change return types to deal with upper bounds

### DISCUSSION THREAD

**Discussion by boanergies** [Textualize/textual#2342] [Q&A] [answered]
Have you checked closed issues? https://github.com/Textualize/textual/issues?q=is%3Aissue+is%3Aclosed

When using code like the example below and trying to run mypy on it

```
from textual.app import App, ComposeResult
from textual.widgets import OptionList


class ExampleApp(App):
    def compose(self) -> ComposeResult:
        yield OptionList("cool", "cool", id="watch")
        yield OptionList("to clear", id="update")

    async def on_option_list_option_selected(self, message: OptionList.OptionSelected):
        if message.option_list.id == "watch":
            option = self.query_one("#update")
            option.clear_options()


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```
you will then get this error:

```
example.py:13: error: "Widget" has no attribute "clear_options"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

This is because the return type for `query_one` should be a [type with an upper bound](https://mypy.readthedocs.io/en/stable/generics.html#type-variables-with-upper-bounds). Another example explaining this type hint can be found [here](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-type-of-class-objects)

It's probably not an immediate issue but it will make using mypy in textual projects better

**Comment by davep** [comment, marked as answer]
This would be the perfect time to use the [`expect_type` parameter of `query_one`](https://textual.textualize.io/api/dom_node/#textual.dom.DOMNode.query_one), for example:

```python
option = self.query_one("#update", OptionList)
```

**Reply by boanergies** [reply]
Does using that argument somehow edit the return type of the type hint?

**Reply by davep** [reply]
Yes, [see `query_one` in the source for he detail](https://github.com/Textualize/textual/blob/cab4925eaa75e3fe2273ea64b667db1e934a76f0/src/textual/dom.py#L1054). Both this:

```python
found = self.query_one(SomeWidget)
```

and

```python
found = self.query_one("#some-query-here", SomeWidget)
```

will result in `found` being typed correctly.
